functions/webhooks/twitter/TwitterToFlex.ts

```diff
@@ -0,0 +1,270 @@
+};
+
+const twitterUniqueNamePrefix = 'twitter:';
+
+/**
+ * Validates that the payload is signed with TWITTER_CONSUMER_SECRET so we know it's comming from Twitter
+ */
+const isValidTwitterPayload = (event: Body, consumerSecret: string): boolean => {
+  if (!event.bodyAsString || !event.xTwitterWebhooksSignature) return false;
+
+  const expectedSignature = crypto
+    .createHmac('sha256', consumerSecret)
+    .update(event.bodyAsString)
+    .digest('base64');
+
+  const isValidRequest = crypto.timingSafeEqual(
+    Buffer.from(event.xTwitterWebhooksSignature),
+    Buffer.from(`sha256=${expectedSignature}`),
+  );
+
+  return isValidRequest;
+};
+
+/**
+ * Retrieves a channel by sid or uniqueName
+ */
+const retrieveChannel = (context: Context<EnvVars>, uniqueName: string) =>
+  context
+    .getTwilioClient()
+    .chat.services(context.CHAT_SERVICE_SID)
+    .channels(uniqueName)
+    .fetch();
+
+/**
+ * Creates a new Flex channel in the Twitter Flex Flow and subscribes webhooks to it's events
+ */
+const createTwitterChannel = async (
+  context: Context<EnvVars>,
+  uniqueChannelName: string,
+  senderId: string,
+  uniqueSenderName: string, // Unique identifier for this person
+  senderScreenName: string, // Twiter handle to show friendly info
+  forUserId: string,
+) => {
+  const twilioNumber = `${twitterUniqueNamePrefix}${forUserId}`;
+
+  const client = context.getTwilioClient();
+
+  const channel = await client.flexApi.channel.create({
+    flexFlowSid: context.TWITTER_FLEX_FLOW_SID,
+    identity: uniqueSenderName,
+    chatUserFriendlyName: senderScreenName,
+    chatFriendlyName: uniqueChannelName,
+    chatUniqueName: uniqueChannelName,
+    target: uniqueChannelName,
+  });
+
+  const channelAttributes = JSON.parse(
+    (
+      await client.chat
+        .services(context.CHAT_SERVICE_SID)
+        .channels(channel.sid)
+        .fetch()
+    ).attributes,
+  );
+
+  await client.chat
+    .services(context.CHAT_SERVICE_SID)
+    .channels(channel.sid)
+    .update({
+      attributes: JSON.stringify({
+        ...channelAttributes,
+        channel_type: 'twitter',
+        twitterUserHandle: senderScreenName,
+        // customChannelType: 'twitter',
+        twilioNumber,
+      }),
+    });
+
+  /* const onMessageSent = */
+  await client.chat
+    .services(context.CHAT_SERVICE_SID)
+    .channels(channel.sid)
+    .webhooks.create({
+      type: 'webhook',
+      configuration: {
+        method: 'POST',
+        url: `https://${context.DOMAIN_NAME}/webhooks/twitter/FlexToTwitter?recipientId=${senderId}`,
+        filters: ['onMessageSent'],
+      },
+    });
+
+  /* const onChannelUpdated = */
+  await client.chat
+    .services(context.CHAT_SERVICE_SID)
+    .channels(channel.sid)
+    .webhooks.create({
+      type: 'webhook',
+      configuration: {
+        method: 'POST',
+        url: `https://${context.DOMAIN_NAME}/webhooks/twitter/FlexChannelUpdate`,
+        filters: ['onChannelUpdated'],
```

Comment:
Who calls this function? Is it called only when we need to remove a Flex channel?

Reply:
It is called on every channel update. In particular we are now interested when this channel is marked as `INACTIVE` to perform the appropriate cleanup.